sklearn/partial_dependence.py

```diff
@@ -0,0 +1,574 @@
+
+    target_variables = np.asarray(target_variables, dtype=np.int32,
+                                  order='C').ravel()
+
+    if any([not (0 <= fx < n_features) for fx in target_variables]):
+        raise ValueError('target_variables must be in [0, %d]'
+                         % (n_features - 1))
+
+    if X is not None:
+        X = check_array(X, dtype=DTYPE, order='C')
+        grid, axes = _grid_from_X(X[:, target_variables], percentiles,
+                                  grid_resolution)
+    else:
+        assert grid is not None
+        # don't return axes if grid is given
+        axes = None
+        # grid must be 2d
+        if grid.ndim == 1:
+            grid = grid[:, np.newaxis]
+        if grid.ndim != 2:
+            raise ValueError('grid must be 2d but is %dd' % grid.ndim)
+
+    grid = np.asarray(grid, dtype=DTYPE, order='C')
+    assert grid.shape[1] == target_variables.shape[0]
+
+    if method == 'recursion':
+        if isinstance(est, BaseGradientBoosting):
+            n_trees_per_stage = est.estimators_.shape[1]
+            n_estimators = est.estimators_.shape[0]
+            learning_rate = est.learning_rate
+        else:
+            n_trees_per_stage = 1
+            n_estimators = len(est.estimators_)
+            learning_rate = 1.
+        pdp = np.zeros((n_trees_per_stage, grid.shape[0],), dtype=np.float64,
+                       order='C')
+        for stage in range(n_estimators):
+            for k in range(n_trees_per_stage):
+                if isinstance(est, BaseGradientBoosting):
+                    tree = est.estimators_[stage, k].tree_
+                else:
+                    tree = est.estimators_[stage].tree_
+                _partial_dependence_tree(tree, grid, target_variables,
+                                         learning_rate, pdp[k])
+        if isinstance(est, ForestRegressor):
+            pdp /= n_estimators
+    elif method == 'exact':
+        pdp = _exact_partial_dependence(est, target_variables, grid, X)
+    elif method == 'estimated':
+        pdp = _estimated_partial_dependence(est, target_variables, grid, X)
+    else:
+        raise ValueError('method "%s" is invalid. Use "recursion", "exact", '
+                         '"estimated", or None.' % method)
+
+    return pdp, axes
+
+
+def plot_partial_dependence(gbrt, X, features, feature_names=None,
+                            label=None, n_cols=3, grid_resolution=100,
+                            percentiles=(0.05, 0.95), n_jobs=1,
+                            verbose=0, ax=None, line_kw=None,
+                            contour_kw=None, **fig_kw):
+    """Partial dependence plots for ``features``.
+
+    The ``len(features)`` plots are arranged in a grid with ``n_cols``
+    columns. Two-way partial dependence plots are plotted as contour
+    plots.
+
+    Read more in the :ref:`User Guide <partial_dependence>`.
+
+    Parameters
+    ----------
+    gbrt : BaseGradientBoosting
```

Comment:
`plot_partial_dependence` is on my TODO. It is untouched from the existing one in `ensemble`